File: algorithm/Python3_version/leetcode135.candy.py
class Solution:
    def candy(self, ratings: List[int]) -> int:
        n = len(ratings)
        dp = [1]*n
        for i in range(1, n):
            if ratings[i] > ratings[i-1]:
                dp[i] = dp[i-1]+1
        sum = dp[n-1]
        for i in range(n-2, -1, -1):
            if ratings[i] > ratings[i+1]:
                dp[i] = max(dp[i], dp[i+1]+1)
            sum += dp[i]
        return sum

File: algorithm/leetcode135.candy.cpp
class Solution {
public:
    int candy(vector<int>& ratings) {
    	int i, j, n = ratings.size(), sum = 0;
    	vector<int> dp(n,1);
    	for(i = 1, j = n-2; i < n && j >= 0; ++i,--j)
    	{
    		if(ratings[i] > ratings[i-1])
    			dp[i] = max(dp[i], dp[i-1]+1);
            if(ratings[j] > ratings[j+1])
                dp[j] = max(dp[j], dp[j+1]+1);
    	}
        for(i = 0; i < n; ++i)
            sum += dp[i];
    	return sum;
    }
};